File: src/main/services/proxy/usageParser.ts
/**
 * Usage Parser Service
 * Parses token usage from API responses (Claude and OpenAI formats)
 */

// Token usage from API response
export interface TokenUsage {
  inputTokens: number
  outputTokens: number
  cacheReadTokens?: number
  cacheCreationTokens?: number
}

// Claude API response usage format
interface ClaudeUsage {
  input_tokens?: number
  output_tokens?: number
  cache_read_input_tokens?: number
  cache_creation_input_tokens?: number
}

// OpenAI API response usage format
interface OpenAIUsage {
  prompt_tokens?: number
  completion_tokens?: number
  total_tokens?: number
}

// Generic API response with usage
interface ApiResponse {
  usage?: ClaudeUsage | OpenAIUsage
  model?: string
}

/**
 * Parse token usage from API response
 * Supports both Claude and OpenAI response formats
 */
export function parseUsageFromResponse(response: unknown): TokenUsage | null {
  if (!response || typeof response !== 'object') {
    return null
  }

  const resp = response as ApiResponse

  if (!resp.usage) {
    return null
  }

  const usage = resp.usage

  // Try Claude format first
  if ('input_tokens' in usage || 'output_tokens' in usage) {
    const claudeUsage = usage as ClaudeUsage
    return {
      inputTokens: claudeUsage.input_tokens || 0,
      outputTokens: claudeUsage.output_tokens || 0,
      cacheReadTokens: claudeUsage.cache_read_input_tokens || 0,
      cacheCreationTokens: claudeUsage.cache_creation_input_tokens || 0,
    }
  }

  // Try OpenAI format
  if ('prompt_tokens' in usage || 'completion_tokens' in usage) {
    const openaiUsage = usage as OpenAIUsage
    return {
      inputTokens: openaiUsage.prompt_tokens || 0,
      outputTokens: openaiUsage.completion_tokens || 0,
      cacheReadTokens: 0,
      cacheCreationTokens: 0,
    }
  }

  return null
}

/**
 * Parse model name from API response
 */
export function parseModelFromResponse(response: unknown): string | null {
  if (!response || typeof response !== 'object') {
    return null
  }

  const resp = response as ApiResponse
  return resp.model || null
}

/**
 * Parse usage from streaming response
 * For streaming, usage is typically in the last chunk with type 'message_delta' or 'message_stop'
 */
export function parseUsageFromStreamChunk(chunk: string): TokenUsage | null {
  try {
    // Handle SSE format: "data: {...}"
    const lines = chunk.split('\n').filter(line => line.startsWith('data: '))

    for (const line of lines) {
      const jsonStr = line.slice(6) // Remove "data: " prefix

      if (jsonStr === '[DONE]') {
        continue
      }

      try {
        const data = JSON.parse(jsonStr)

        // Claude streaming format - look for message_delta with usage
        if (data.type === 'message_delta' && data.usage) {
          return {
            inputTokens: data.usage.input_tokens || 0,
            outputTokens: data.usage.output_tokens || 0,
            cacheReadTokens: data.usage.cache_read_input_tokens || 0,
            cacheCreationTokens: data.usage.cache_creation_input_tokens || 0,
          }
        }

        // Claude streaming - message_start contains input token count
        if (data.type === 'message_start' && data.message?.usage) {
          return {
            inputTokens: data.message.usage.input_tokens || 0,
            outputTokens: 0,
            cacheReadTokens: data.message.usage.cache_read_input_tokens || 0,
            cacheCreationTokens: data.message.usage.cache_creation_input_tokens || 0,
          }
        }

        // OpenAI streaming format - usage in final chunk
        if (data.usage) {
          return parseUsageFromResponse(data)
        }
      } catch {
        // Skip invalid JSON lines
        continue
      }
    }
  } catch {
    // Ignore parsing errors
  }

  return null
}

/**
 * Accumulate usage from multiple stream chunks
 * Claude sends input tokens in message_start and output tokens in message_delta
 */
export class StreamUsageAccumulator {
  private inputTokens = 0
  private outputTokens = 0
  private cacheReadTokens = 0
  private cacheCreationTokens = 0
  private model: string | null = null

  processChunk(chunk: string): void {
    try {
      const lines = chunk.split('\n').filter(line => line.startsWith('data: '))

      for (const line of lines) {
        const jsonStr = line.slice(6)

        if (jsonStr === '[DONE]') {
          continue
        }

        try {
          const data = JSON.parse(jsonStr)

          // Extract model from message_start
          if (data.type === 'message_start' && data.message?.model) {
            this.model = data.message.model
          }

          // Claude message_start - input tokens
          if (data.type === 'message_start' && data.message?.usage) {
            this.inputTokens = data.message.usage.input_tokens || 0
            this.cacheReadTokens = data.message.usage.cache_read_input_tokens || 0
            this.cacheCreationTokens = data.message.usage.cache_creation_input_tokens || 0
          }

          // Claude message_delta - output tokens
          if (data.type === 'message_delta' && data.usage) {
            this.outputTokens = data.usage.output_tokens || 0
          }

          // OpenAI format - full usage in final chunk
          if (data.usage && ('prompt_tokens' in data.usage || 'completion_tokens' in data.usage)) {
            this.inputTokens = data.usage.prompt_tokens || 0
            this.outputTokens = data.usage.completion_tokens || 0
          }

          // Extract model from OpenAI response
          if (data.model && !this.model) {
            this.model = data.model
          }
        } catch {
          continue
        }
      }
    } catch {
      // Ignore errors
    }
  }

  getUsage(): TokenUsage | null {
    if (this.inputTokens === 0 && this.outputTokens === 0) {
      return null
    }

    return {
      inputTokens: this.inputTokens,
      outputTokens: this.outputTokens,
      cacheReadTokens: this.cacheReadTokens,
      cacheCreationTokens: this.cacheCreationTokens,
    }
  }

  getModel(): string | null {
    return this.model
  }

  reset(): void {
    this.inputTokens = 0
    this.outputTokens = 0
    this.cacheReadTokens = 0
    this.cacheCreationTokens = 0
    this.model = null
  }
}

File: src-tauri/src/proxy/usage_parser.rs
use serde::{Deserialize, Serialize};

/// Token usage from API response
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TokenUsage {
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_read_tokens: i64,
    pub cache_creation_tokens: i64,
}

/// Parse usage from a non-streaming JSON response
pub fn parse_usage_from_response(body: &str) -> (Option<TokenUsage>, Option<String>) {
    let json: serde_json::Value = match serde_json::from_str(body) {
        Ok(v) => v,
        Err(_) => return (None, None),
    };

    let model = json
        .get("model")
        .or_else(|| json.get("response").and_then(|r| r.get("model")))
        .and_then(|m| m.as_str())
        .map(|s| s.to_string());
    let usage = parse_usage_from_json(&json);

    (usage, model)
}

fn parse_usage_from_json(json: &serde_json::Value) -> Option<TokenUsage> {
    let usage = json
        .get("usage")
        .or_else(|| json.get("response").and_then(|r| r.get("usage")))
        .or_else(|| find_usage_value(json))?;
    parse_usage_value(usage)
}

fn find_usage_value(json: &serde_json::Value) -> Option<&serde_json::Value> {
    match json {
        serde_json::Value::Object(map) => {
            if let Some(usage) = map.get("usage") {
                if usage.get("input_tokens").is_some()
                    || usage.get("output_tokens").is_some()
                    || usage.get("prompt_tokens").is_some()
                    || usage.get("completion_tokens").is_some()
                {
                    return Some(usage);
                }
            }
            map.values().find_map(find_usage_value)
        }
        serde_json::Value::Array(items) => items.iter().find_map(find_usage_value),
        _ => None,
    }
}

fn parse_usage_value(usage: &serde_json::Value) -> Option<TokenUsage> {
    let cache_read_tokens = usage
        .get("cache_read_input_tokens")
        .or_else(|| {
            usage
                .get("input_tokens_details")
                .and_then(|details| details.get("cached_tokens"))
        })
        .and_then(|v| v.as_i64())
        .unwrap_or(0);

    // Claude / OpenAI Responses format
    if usage.get("input_tokens").is_some() || usage.get("output_tokens").is_some() {
        return Some(TokenUsage {
            input_tokens: usage
                .get("input_tokens")
                .and_then(|v| v.as_i64())
                .unwrap_or(0),
            output_tokens: usage
                .get("output_tokens")
                .and_then(|v| v.as_i64())
                .unwrap_or(0),
            cache_read_tokens,
            cache_creation_tokens: usage
                .get("cache_creation_input_tokens")
                .and_then(|v| v.as_i64())
                .unwrap_or(0),
        });
    }

    // OpenAI format
    if usage.get("prompt_tokens").is_some() || usage.get("completion_tokens").is_some() {
        return Some(TokenUsage {
            input_tokens: usage
                .get("prompt_tokens")
                .and_then(|v| v.as_i64())
                .unwrap_or(0),
            output_tokens: usage
                .get("completion_tokens")
                .and_then(|v| v.as_i64())
                .unwrap_or(0),
            cache_read_tokens: 0,
            cache_creation_tokens: 0,
        });
    }

    None
}

/// Accumulate usage from SSE streaming chunks.
///
/// Handles cross-chunk line splitting by buffering incomplete lines.
pub struct StreamUsageAccumulator {
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_read_tokens: i64,
    pub cache_creation_tokens: i64,
    pub model: Option<String>,
    data_event_count: usize,
    invalid_json_count: usize,
    event_types: Vec<String>,
    top_level_keys: Vec<String>,
    /// Buffer for incomplete lines that span across TCP chunks
    line_buffer: String,
}

impl StreamUsageAccumulator {
    pub fn new() -> Self {
        Self {
            input_tokens: 0,
            output_tokens: 0,
            cache_read_tokens: 0,
            cache_creation_tokens: 0,
            model: None,
            data_event_count: 0,
            invalid_json_count: 0,
            event_types: Vec::new(),
            top_level_keys: Vec::new(),
            line_buffer: String::new(),
        }
    }

    pub fn process_chunk(&mut self, chunk: &str) {
        // Prepend any buffered incomplete line from previous chunk
        let data_to_process = if self.line_buffer.is_empty() {
            chunk.to_string()
        } else {
            let combined = format!("{}{}", self.line_buffer, chunk);
            self.line_buffer.clear();
            combined
        };

        // If the chunk doesn't end with a newline, the last line is incomplete
        let ends_with_newline = data_to_process.ends_with('\n');
        let lines: Vec<&str> = data_to_process.lines().collect();

        for (i, line) in lines.iter().enumerate() {
            // If this is the last line and chunk didn't end with newline,
            // buffer it for the next chunk
            if i == lines.len() - 1 && !ends_with_newline {
                self.line_buffer = line.to_string();
                break;
            }

            self.process_sse_line(line);
        }
    }

    fn process_sse_line(&mut self, line: &str) {
        if !line.starts_with("data: ") {
            return;
        }
        let json_str = &line[6..];
        if json_str == "[DONE]" {
            return;
        }
        self.data_event_count += 1;

        let data: serde_json::Value = match serde_json::from_str(json_str) {
            Ok(v) => v,
            Err(_) => {
                self.invalid_json_count += 1;
                return;
            }
        };

        if let Some(obj) = data.as_object() {
            for key in obj.keys() {
                push_limited_unique(&mut self.top_level_keys, key);
            }
        }

        // Extract model
        if let Some(msg) = data.get("message") {
            if let Some(m) = msg.get("model").and_then(|v| v.as_str()) {
                self.model = Some(m.to_string());
            }
        }
        if self.model.is_none() {
            if let Some(m) = data.get("model").and_then(|v| v.as_str()) {
                self.model = Some(m.to_string());
            }
        }

        let event_type = data.get("type").and_then(|v| v.as_str());
        if let Some(event_type) = event_type {
            push_limited_unique(&mut self.event_types, event_type);
        }

        // Claude message_start — input tokens
        if event_type == Some("message_start") {
            if let Some(usage) = data.get("message").and_then(|m| m.get("usage")) {
                self.input_tokens = usage
                    .get("input_tokens")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0);
                self.cache_read_tokens = usage
                    .get("cache_read_input_tokens")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0);
                self.cache_creation_tokens = usage
                    .get("cache_creation_input_tokens")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0);
            }
        }

        // Claude message_delta — output tokens (final cumulative value)
        if event_type == Some("message_delta") {
            if let Some(usage) = data.get("usage") {
                if let Some(v) = usage.get("output_tokens").and_then(|v| v.as_i64()) {
                    self.output_tokens = v;
                }
            }
        }

        // OpenAI format — full usage in final chunk
        if let Some(usage) = data.get("usage") {
            if usage.get("prompt_tokens").is_some() {
                self.input_tokens = usage
                    .get("prompt_tokens")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0);
                self.output_tokens = usage
                    .get("completion_tokens")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0);
            }
        }

        // OpenAI Responses streaming format — final events often wrap usage
        // under `response.usage` instead of putting it at the top level.
        if let Some(usage) = data
            .get("response")
            .and_then(|response| response.get("usage"))
            .or_else(|| {
                if event_type
                    .map(|event_type| event_type.starts_with("response."))
                    .unwrap_or(false)
                {
                    find_usage_value(&data)
                } else {
                    None
                }
            })
        {
            if let Some(parsed) = parse_usage_value(usage) {
                self.input_tokens = parsed.input_tokens;
                self.output_tokens = parsed.output_tokens;
                self.cache_read_tokens = parsed.cache_read_tokens;
                self.cache_creation_tokens = parsed.cache_creation_tokens;
            }
        }
        if self.model.is_none() {
            if let Some(m) = data
                .get("response")
                .and_then(|response| response.get("model"))
                .and_then(|v| v.as_str())
            {
                self.model = Some(m.to_string());
            }
        }
    }

    /// Flush any remaining buffered line (call when stream ends)
    pub fn flush(&mut self) {
        if !self.line_buffer.is_empty() {
            let line = std::mem::take(&mut self.line_buffer);
            self.process_sse_line(&line);
        }
    }

    pub fn get_usage(&self) -> Option<TokenUsage> {
        if self.input_tokens == 0 && self.output_tokens == 0 {
            return None;
        }
        Some(TokenUsage {
            input_tokens: self.input_tokens,
            output_tokens: self.output_tokens,
            cache_read_tokens: self.cache_read_tokens,
            cache_creation_tokens: self.cache_creation_tokens,
        })
    }

    pub fn diagnostics_summary(&self) -> String {
        format!(
            "data_events={}, invalid_json={}, event_types=[{}], top_level_keys=[{}], model_seen={}",
            self.data_event_count,
            self.invalid_json_count,
            self.event_types.join(","),
            self.top_level_keys.join(","),
            self.model.as_deref().unwrap_or("")
        )
    }
}

fn push_limited_unique(values: &mut Vec<String>, value: &str) {
    if values.len() >= 16 || values.iter().any(|existing| existing == value) {
        return;
    }
    values.push(value.to_string());
}

/// Parse usage from collected response data (handles both SSE and JSON)
pub fn parse_usage_from_response_data(
    response_text: &str,
    content_type: &str,
) -> (Option<TokenUsage>, Option<String>, bool) {
    let is_sse = content_type.contains("text/event-stream")
        || response_text.trim_start().starts_with("data: ");

    if is_sse {
        let mut acc = StreamUsageAccumulator::new();
        acc.process_chunk(response_text);
        acc.flush();
        (acc.get_usage(), acc.model.clone(), true)
    } else {
        let (usage, model) = parse_usage_from_response(response_text);
        (usage, model, false)
    }
}
